keps/sig-node/20200205-build-kubelet-without-docker.md

```diff
@@ -0,0 +1,294 @@
+code/dependency's removal.
+
+A prototype is available in [kubernetes/kubernetes#87746](https://github.com/kubernetes/kubernetes/pull/87746).
+
+To ensure that this dockerless Kubelet continues to function we will add CI building in this mode,
+and CI running end to end tests against it (to be elaborated on in the test plan). Additionally, to
+ensure the Kubelet doesn't introduce new dependencies on the `docker/docker`
+Golang library, we will add automated tooling enforcing that only the
+`dockershim` can depend on `docker/docker`.
+
+One quick additional note - currently `cadvisor` also depends on the
+`docker/docker` client library. Since `kubelet` depends on `cadvisor`, the
+`kubelet` can not truly be rid of the `docker/docker` client library until `cadvisor` no
+longer depends on `docker/docker`. Work to remove the `docker/docker` dependency
+from `cadvisor` is being dealt with in separate workstreams.
+
+This proposal follows the previous patterns for similar work, namely the efforts
+of sig-cloud-provider to support [Building Without In-Tree Cloud Providers](/keps/sig-cloud-provider/20190729-building-without-in-tree-providers.md).
+
+### User Stories
+
+#### Story 1
+
+As a developer working to make Docker function like any other remote container
+runtime, I am attempting to validate that my proposed solution
+functions correctly. Using this dockerless build ensures the Kubelet contains no
+"Docker-specific" code, and any success/failures can be attributed to my
+implementation.
+
+### Implementation Details/Notes/Constraints
+
+A couple high level notes (to be extended over time):
+
+- We implement this functionality using a synthetic `dockerless` tag in go build
+  constraints on relevant sources. If `GOFLAGS=-tags=dockerless` during build,
+  then "Docker-specific" code will be excluded from the Kubelet build. With no
+  "Docker-specific" code, we should also be excluding the dependency on
+  `docker/docker`.
+
+### Risks and Mitigations
+
+This feature is only developer facing, which removes a large class of risks.
+
+The largest remaining risk is that the build tags fall out of date, or are
+burdensome to continue updating, which leads to the dockerless Kubelet build
+breaking and/or being costly to maintain. This risk grows the longer the
+dockerless Kubelet exists (i.e. the longer it takes to move `dockershim` out of
+tree/have Docker support the CRI). Fortunately, these risks can be mitigated via
+the CI tooling discussed earlier.
+
+## Design Details
+
+### Test Plan
+
+**Note:** *Section not required until targeted at a release.*
+
+We envision the following testing:
+
+1. A verification, run during pre-submit, which ensures the Kubelet builds when the `dockerless` tag is
+   enabled.
+2. A verification, run during pre-submit, which ensures that only the imports of
+   `github.com/docker/docker` occur within `pkg/kubelet/dockershim`.
+3. Unit tests, run during pre-submit, which execute the standard `pkg/kubelet/...`
```

Comment:
@derekwaynecarr when you talked about a smoke test, I wasn't sure if you were suggesting this, or if #1 is what you were referring to?

If we do want to run the unit tests with the `dockerless` tag enabled... I'm not actually positive how to set up CI to do so.
Not something we need to sort out this second :) 

For local testing, I think `go test --tags=dockerless ./pkg/kubelet/...` works. Although I imagine we will want to use `bazel` in the actual CI... anyway, we can finalize the details when the time comes.

Reply:
yep, we will work through it Matt

Reply:
agree with @dims we can work through it during implementation.